shell/platform/android/android_image_generator.cc

```diff
@@ -0,0 +1,193 @@
+
+#include "flutter/shell/platform/android/android_image_generator.h"
+
+#include <android/bitmap.h>
+
+#include "flutter/fml/platform/android/jni_util.h"
+
+namespace flutter {
+
+static fml::jni::ScopedJavaGlobalRef<jclass>* g_flutter_jni_class = nullptr;
+static jmethodID g_decode_image_method = nullptr;
+
+AndroidImageGenerator::~AndroidImageGenerator() = default;
+
+AndroidImageGenerator::AndroidImageGenerator(sk_sp<SkData> data)
+    : data_(data), image_info_(SkImageInfo::MakeUnknown()) {}
+
+const SkImageInfo& AndroidImageGenerator::GetInfo() {
+  header_decoded_latch_.Wait();
+  return image_info_;
+}
+
+unsigned int AndroidImageGenerator::GetFrameCount() const {
+  return 1;
+}
+
+unsigned int AndroidImageGenerator::GetPlayCount() const {
+  return 1;
+}
+
+const ImageGenerator::FrameInfo AndroidImageGenerator::GetFrameInfo(
+    unsigned int frame_index) const {
+  return {.required_frame = std::nullopt,
+          .duration = 0,
+          .disposal_method = SkCodecAnimation::DisposalMethod::kKeep};
+}
+
+SkISize AndroidImageGenerator::GetScaledDimensions(float desired_scale) {
+  return GetInfo().dimensions();
+}
+
+bool AndroidImageGenerator::GetPixels(const SkImageInfo& info,
+                                      void* pixels,
+                                      size_t row_bytes,
+                                      unsigned int frame_index,
+                                      std::optional<unsigned int> prior_frame) {
+  fully_decoded_latch_.Wait();
+
+  if (kRGBA_8888_SkColorType != info.colorType()) {
+    return false;
+  }
+
+  switch (info.alphaType()) {
+    case kOpaque_SkAlphaType:
+      if (kOpaque_SkAlphaType != GetInfo().alphaType()) {
+        return false;
+      }
+      break;
+    case kPremul_SkAlphaType:
+      break;
+    default:
+      return false;
+  }
+
+  // Since this decoder is only installed for API 28 and 29, there is no
+  // supported way to work around this unfortunate copy. It's possible to
+  // produce an `SkImage` direcly from an `AHardwareBuffer`, but
+  // `AndroidBitmap_getHardwareBuffer` and `Bitmap.getHardwareBuffer` are only
+  // available in API 30+ and 31+ respectively.
```

Comment:
There is! But we'd have to do some dynamic lookups to see whether symbols are available.

I'm ok having this in place without doing that yet, but we should file an issue if one doesn't exist and add a TODO here for it.

Reply:
I got most of the way through resolving/calling `AndroidBitmap_getHardwareBuffer` at runtime, but even [SkImage::MakeFromAHardwareBuffer](https://github.com/google/skia/blob/1f193df9b393d50da39570dab77a0bb5d28ec8ef/src/image/SkImage_Gpu.cpp#L688-L694) and the underlying [GrAHardwareBufferImageGenerator](https://github.com/google/skia/blob/1f193df9b393d50da39570dab77a0bb5d28ec8ef/src/gpu/GrAHardwareBufferImageGenerator.cpp) are preprocessed out with with `__ANDROID_API__ >= 26` due to reliance on e.g. `AHardwareBuffer_acquire`... :(

I could probably propose a reasonable patch to cheaply resolve the symbols in Skia instead of gating like this, but we gotta make some behavior decisions/maybe provide a way to query support, and so I made a note to discuss with Skia friends and follow up.

Reply:
@LeonScroggins

Reply:
I think it would make sense to update Skia to check for the symbols rather than using the prepocessor to remove them entirely.

Reply:
I think for now we should just add a TODO to fix this when Skia API is available.